improver/calibration/rainforest_calibration.py

```diff
@@ -491,200 +488,101 @@ def _calculate_error_probabilities(
-            the expected bounds of the forecast parameter can arise. These values occur when
-            when the input forecast value is between error thresholds and there exists a
-            lower bound on the observable value (eg. 0 in the case of rainfall).
-
-            In this situation, error thresholds below the residual value (min(obs) - fcst)
-            must have a probability of exceedance of 1, whereas as error thresholds above
-            this value can take on any value between [0, 1]. In the subsequent step where
-            error percentile values are extracted, the linear interpolation in mapping from
-            probabilities to percentiles can give percentile values that lie below the
-            residual value; when these are applied to the forecast value, they result in
-            forecast values outside the expected bounds of the forecast parameter in the
-            resultant sub-ensemble.
-
-            To address this, we remap all values outside of the expected bounds to nearest
-            bound (eg. negative values are mapped to 0 in the case of rainfall).
+        """
+        Add the error CDF to the NWP forecast and extract probabilities at output thresholds
+        for all realizations.
 
         Args:
-            forecast_cube:
-                Cube containing the forecast to be calibrated.
-            error_percentiles_cube:
-                Cube containing percentile values for the error distributions.
+            error_CDF:
+                Cube containing the CDF of probabilities for each enemble member at error
+                threhsolds.
+            forecast:
+                Cube containing NWP ensemble forecast.
+            output_thresholds:
+                Ordered list of thresholds at which to calculate the output probabilities.
 
         Returns:
-            Cube containing the forecast sub-ensembles.
+            Cube containing probabilities at output thresholds for all realizations. Dimensions
+            are same as forecast cube with additional threshold dimension first.
         """
-        # Apply the error_percentiles to the forecast_cube (additive correction)
-        forecast_subensembles_data = (
-            forecast_cube.data[:, np.newaxis] + error_percentiles_cube.data
+
+        # Add error to forecast
+        error_values = error_CDF.coord(
+            "forecast_error_of_lwe_thickness_of_precipitation_amount"
+        ).points
+        num_forecast_dims = len(forecast.coords(dim_coords=True))
+        for i in range(num_forecast_dims):
+            error_values = np.expand_dims(error_values, axis=-1)
+        thresholds = error_values + np.expand_dims(forecast.data, axis=0)
+        probabilities = error_CDF.data
+
+        # Add bounds
+        epsilon = 0.0001
+        thresholds = np.concatenate(
+            [thresholds[[0]] - epsilon, thresholds, thresholds[[-1]] + epsilon], axis=0,
```

Comment:
I've changed this to do something similar to the expected value cli, using the floating-point epsilon. I'm now using `get_bounds_of_distribution` (also changed in `_evaluate_probabilities`). I don't think `insert_lower_and_upper_endpoint_to_1d_array` will work here since `thresholds` is not a 1-d array.

Reply:
No worries. Thanks for doing. As to the `eps` approach, I can understand the reasoning however note that if the `error_thresholds` don't provide suitable bounding values, then it may be possible that the probabilities are interpolated over a VERY small gap. That being said, I appreciate that specifying a bound (such as upper ECC bound) may result in interpolations over a very large gap. I'm not sure there is a good solution here other than to say that the user should ensure they have suitable limits for the upper and lower `error_thresholds`.